packages/fx-core/src/component/driver/aad/create.ts

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { StepDriver } from "../../interface/stepDriver";
+import { DriverContext } from "../../interface/commonArgs";
+import { Service } from "typedi";
+import { CreateAadAppArgs } from "./interface/createAadAppArgs";
+import { AadAppClient } from "./utility/aadAppClient";
+import { CreateAadAppOutput } from "./interface/createAadAppOutput";
+import { M365TokenProvider, SystemError, UserError } from "@microsoft/teamsfx-api";
+import { GraphScopes } from "../../../common/tools";
+import { Constants } from "../../resource/aadApp/constants";
+import { InvalidParameterUserError } from "./error/invalidParameterUserError";
+import { MissingEnvUserError } from "./error/missingEnvError";
+import { UnhandledSystemError, UnhandledUserError } from "./error/unhandledError";
+import axios from "axios";
+
+const actionName = "aadApp/create"; // DO NOT MODIFY the name
+const helpLink = "https://aka.ms/teamsfx-actions/aadapp-create";
+const driverConstants = {
+  generateSecretErrorMessageKey: "driver.aadApp.error.generateSecretFailed",
+};
+
+@Service(actionName) // DO NOT MODIFY the service name
+export class CreateAadAppDriver implements StepDriver {
+  public async run(args: CreateAadAppArgs, context: DriverContext): Promise<Map<string, string>> {
+    try {
+      this.validateArgs(args);
+      const aadAppClient = new AadAppClient(context.m365TokenProvider);
+      const aadAppState = this.loadCurrentState();
+      if (!aadAppState.AAD_APP_CLIENT_ID) {
+        // Create new AAD app if no client id exists
+        const aadApp = await aadAppClient.createAadApp(args.name);
+        aadAppState.AAD_APP_CLIENT_ID = aadApp.appId!;
+        aadAppState.AAD_APP_OBJECT_ID = aadApp.id!;
+        await this.setAadEndpointInfo(context.m365TokenProvider, aadAppState);
+      }
+
+      if (args.generateClientSecret && !aadAppState.SECRET_AAD_APP_CLIENT_SECRET) {
+        // Create new client secret if no client secret exists
+        if (!aadAppState.AAD_APP_OBJECT_ID) {
```

Comment:
So if it is the existing aad app scenario and would like to generate secret, object id must be provided?

Reply:
Yes, there may be cases that user does not have permission to query object id by client id. So this action follows existing AAD plugin design to ask user provide object id for existing AAD app.